### PR TITLE
Addressing APIs review comments [Part1]

### PR DESCRIPTION
**Pull request by msherif1234**
This PR address APIs review comments from https://github.com/openshift/api/pull/1241
Signed-off-by: Mohamed Mahmoud <mmahmoud@redhat.com>

### CODE REVIEW

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewall_types.go:114]: `SourceCIDRs defines the origin of packets that FirewallProtocolRules will be applied to.`

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewall_types.go:118]: `FirewallProtocolRules is a list of per protocol ingress node firewall rules.`

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewall_types.go:126]: `.`

File: api/v1alpha1/ingressnodefirewallconfig_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewallconfig_types.go:27]: Same question here, what's the meaning of ` + `.
`This field is using map[string]string instead of metav1.LabelSelector because it matches the daemonset NodeSelector template. Thus one cannot benefit from additional match criteria that metav1.LabelSelector provides.`

File: api/v1alpha1/ingressnodefirewallconfig_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewallconfig_types.go:37]: `.`

File: api/v1alpha1/ingressnodefirewallnodestate_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewallnodestate_types.go:25]: `.`

File: api/v1alpha1/ingressnodefirewallnodestate_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewallnodestate_types.go:38]: `.`

**Review by andreaskaris — CHANGES_REQUESTED** [review]

File: README.md
Comment (msherif1234) [reply README.md:103]: w/o this change e2e won't run I can undo this change and address in separate PR

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (msherif1234) [reply api/v1alpha1/ingressnodefirewall_types.go:63]: to make it visible only to developer and not shown in the APIs doc

File: api/v1alpha1/ingressnodefirewallconfig_types.go
Comment (msherif1234) [reply api/v1alpha1/ingressnodefirewallconfig_types.go]: same answer just for developer comments no visible in the APIs

File: controllers/ingressnodefirewall_controller_test.go
Comment (martinkennelly) [inline controllers/ingressnodefirewall_controller_test.go:6]: nit format

File: controllers/ingressnodefirewallnodestate_controller.go
Comment (martinkennelly) [inline controllers/ingressnodefirewallnodestate_controller.go:19]: format

File: controllers/ingressnodefirewallnodestate_controller.go
Comment (martinkennelly) [inline controllers/ingressnodefirewallnodestate_controller.go:152]: nit: Theres a instr.fromInt

File: pkg/webhook/webhook.go
Comment (martinkennelly) [inline pkg/webhook/webhook.go:33]: you can delete this line

File: pkg/webhook/webhook_suite_test.go
Comment (martinkennelly) [inline pkg/webhook/webhook_suite_test.go:23]: format

File: test/e2e/ingress-node-firewall/ingress-node-firewall.go
Comment (martinkennelly) [inline test/e2e/ingress-node-firewall/ingress-node-firewall.go:7]: format

**Comment by martinkennelly** [comment]
Mainly nits - you can merge after correction without additional review from me.

**Comment by martinkennelly** [comment]
I thought were doing `+unionDiscriminator` so that when protocol is ICMP or ICMPv6, OpenAPI would enforce that we need the ICMP type and code defined.
If it was TCP/UDP/STCP, it would enforce that ports are defined.

**Comment by martinkennelly** [comment]
This is awesome work by the way! Looking good :)

File: README.md
Comment (andreaskaris) [reply README.md:103]: So I think we should just make the change to the doc as:
~~~
export KUBECONFIG=$(pwd)/hack/kubeconfig
~~~

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (andreaskaris) [reply api/v1alpha1/ingressnodefirewall_types.go:63]: Ah ok! I didn't know that, neat!

File: README.md
Comment (andreaskaris) [inline README.md:104]: thanks :-)

**Comment by andreaskaris** [comment]
I'm ok with the changes as soon as Martin's comments were addressed :-)

**Comment by msherif1234** [comment]
> I thought were doing `+unionDiscriminator` so that when protocol is ICMP or ICMPv6, OpenAPI would enforce that we need the ICMP type and code defined. If it was TCP/UDP/STCP, it would enforce that ports are defined.
> 
> This doesn't need to block this PR, but I thought we were going down that route in the API review, the last time I read it.

I am using `+union` is that not good ?

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (JoelSpeed) [inline api/v1alpha1/ingressnodefirewall_types.go:56]: When using a union, you have to have fields that match the discriminants. So protocolRule and icmpRule aren't valid

You need one for each of `icmp`, `icmpv6`, `tcp`, `udp` and `sctp` (note the fields I'm referring to are the json tags, the go types match the discriminant exactly)

File: README.md
Comment (andreaskaris) [inline README.md:76]: isn't that redundant?
~~~
protocol: TCP  <--
tcp:                  <--
  ports:
~~~

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewall_types.go:26]: defines

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewall_types.go:27]: `If set, this field's value matches against the ICMP / ICMP6 header.`

How do we distinguish between "set" and icmp type = 0 ?

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewall_types.go:38]: same here, how do we distinguish between code 0 and unset? otherwise, we might want to change the description's wording and just say it defaults to 0.

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewall_types.go:34]: If set, this field's value matches against the ICMP / ICMP6 header.

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewall_types.go:57]: Is this needed if we have TCP, UDP, etc. below? That seems redundant

File: api/v1alpha1/ingressnodefirewallconfig_types.go
Comment (andreaskaris) [inline api/v1alpha1/ingressnodefirewallconfig_types.go:30]: // tolerations is a list of tolerations applied to the ingress node firewall controller
// DaemonSet.

File: README.md
Comment (msherif1234) [reply README.md:76]: this is my understanding of how the union can be used based on this doc 
https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (msherif1234) [reply api/v1alpha1/ingressnodefirewall_types.go:27]: type of 0 probably doesn't exists I will use configured instead of set

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (msherif1234) [reply api/v1alpha1/ingressnodefirewall_types.go:57]: this is the unionDiscriminator based on whatthe protocol it bring in the right fmt

File: README.md
Comment (andreaskaris) [reply README.md:76]: nope, absolutely my bad. I wasn't aware of that guide :-)

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (andreaskaris) [reply api/v1alpha1/ingressnodefirewall_types.go]: code 0 + type 0 = ICMP echo reply https://en.wikipedia.org/wiki/Ping_(networking_utility)#Echo_reply :-)

Comment (msherif1234) [reply api/v1alpha1/ingressnodefirewall_types.go]: my bad don't remember them had to check the RFC anyway I reworded it now so hopefully we are good now

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (martinkennelly) [inline api/v1alpha1/ingressnodefirewall_types.go:107]: did you try and get `cidr` working?

File: api/v1alpha1/ingressnodefirewallconfig_types.go
Comment (martinkennelly) [inline api/v1alpha1/ingressnodefirewallconfig_types.go:30]: remove the full stop at the end

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (martinkennelly) [reply api/v1alpha1/ingressnodefirewall_types.go:38]: @msherif1234 How was this resolved?

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (msherif1234) [reply api/v1alpha1/ingressnodefirewall_types.go:107]: no I thought I did remove it from the last diffs ?

File: api/v1alpha1/ingressnodefirewallconfig_types.go
Comment (msherif1234) [reply api/v1alpha1/ingressnodefirewallconfig_types.go:30]: I am sure there will be more comments from Joel so I can defer this change to the follow up PR(s) ok ?

File: api/v1alpha1/ingressnodefirewall_types.go
Comment (msherif1234) [reply api/v1alpha1/ingressnodefirewall_types.go:107]: yeah its removed we can check it later if we can make it work I can get IPv4 to work but not CIDR

File: api/v1alpha1/ingressnodefirewallconfig_types.go
Comment (martinkennelly) [reply api/v1alpha1/ingressnodefirewallconfig_types.go:30]: yep

**Review by martinkennelly — COMMENTED** [review]
/lgtm

File: api/v1alpha1/ingressnodefirewallconfig_types.go
Comment (andreaskaris) [reply api/v1alpha1/ingressnodefirewallconfig_types.go:30]: I don't know what the OpenShift API standard is, but all comments in Golang should be complete phrases and have a full stop at the end of a phrase. 
https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

Comment (andreaskaris) [reply api/v1alpha1/ingressnodefirewallconfig_types.go:30]: It's just that right now it reads ... ;-)  So we'll get something like: `A list of tolerations applied to ingress node firewall controller. daemonsets.`  in the explain
// A list of tolerations applied to ingress node firewall controller.
// daemonsets.

Comment (msherif1234) [reply api/v1alpha1/ingressnodefirewallconfig_types.go:30]: feel free to edit it in any language u like @andreaskaris once its merge :)

**Pull request merged**
